tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,207 @@ public void resolve() {
     typeMap = inferrer.getTypeMap();
   }
 
-  public DAGRequest buildScan(boolean idxScan) {
-    if (idxScan) {
-      return buildIndexScan();
-    } else {
-      return buildTableScan();
-    }
+  private ColumnInfo handleColumn() {
+    return ColumnInfo.newBuilder()
+        .setColumnId(-1)
+        .setPkHandle(true)
+        // We haven't changed the field name in protobuf file, but
+        // we need to set this to true in order to retrieve the handle,
+        // so the name 'setPkHandle' may sounds strange.
+        .build();
   }
 
-  private DAGRequest buildIndexScan() {
+  /**
+   * merge buildIndexScan and buildTableScan since they are very much alike.
+   * DAGRequest for IndexScan should also contain filters and aggregation, so
+   * we can reuse this part of logic.
+   *
+   * DAGRequest is made up of a chain of executors with strict orders:
+   * TableScan/IndexScan > Selection > Aggregation > TopN/Limit
+   * DAGRequest must contain one and only one TableScan or IndexScan.
+   *
+   * @param isIndexScan whether this dagRequest is IndexScan
+   * @return final DAGRequest built
+   */
+  public DAGRequest buildScan(boolean isIndexScan) {
     checkArgument(startTs != 0, "timestamp is 0");
-    if (indexInfo == null) {
-      throw new TiClientInternalException("Index is empty for index scan");
-    }
     DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
     Executor.Builder executorBuilder = Executor.newBuilder();
     IndexScan.Builder indexScanBuilder = IndexScan.newBuilder();
+    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+    // find a column's offset in fields
+    Map<ColumnRef, Integer> colOffsetMap = new HashMap<>();
+    // find a column's position in index
+    Map<TiColumnInfo, Integer> colPosMap = new HashMap<>();
```

Comment:
`colOffsetMap` is a HashMap which indicates the column's offset in dagRequest, while `colPosMap` points its position in index.

Reply:
The two var name is not clear for reading.  Also please make it more clear in comment that how single read is performed in general.

Reply:
Changed, but they are still somewhat confusing. Maybe some more comments are better?